homeassistant/components/rachio/switch.py

```diff
@@ -33,22 +37,31 @@
 ATTR_ZONE_NUMBER = "Zone number"
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+async def async_setup_entry(hass, config_entry, async_add_entities):
     """Set up the Rachio switches."""
     manual_run_time = timedelta(
-        minutes=hass.data[DOMAIN_RACHIO].config.get(CONF_MANUAL_RUN_MINS)
+        minutes=hass.data[DOMAIN_RACHIO][config_entry.entry_id].config.get(
+            CONF_MANUAL_RUN_MINS
+        )
     )
     _LOGGER.info("Rachio run time is %s", str(manual_run_time))
 
     # Add all zones from all controllers as switches
     devices = []
-    for controller in hass.data[DOMAIN_RACHIO].controllers:
-        devices.append(RachioStandbySwitch(hass, controller))
-
-        for zone in controller.list_zones():
-            devices.append(RachioZone(hass, controller, zone, manual_run_time))
-
-    add_entities(devices)
+    for controller in hass.data[DOMAIN_RACHIO][config_entry.entry_id].controllers:
+        devices.append(
+            await hass.async_add_executor_job(RachioStandbySwitch, hass, controller)
```

Comment:
We should avoid jumping often into the executor. Better to create a sync function that does this all and pass that to the executor.

Reply:
fixed